CT log gets into degraded state due to memory pressure

From the Raio2026h2a CT log returned almost exclusively 429 errors (https://issues.chromium.org/issues/434895698#comment14), and we saw similar patches of high errors rates (sometimes manifesting as 500 errors instead) from the Raio2025h2b log. From internal metrics it appears that the errors appears due to the workers isolate reaching memory limits, and the wrangler tail logs confirm this during periods of high error rates.

```
POST https://ct.cloudflare.com/logs/raio2025h2b/ct/v1/add-chain - Exception Thrown @ 9/4/2025, 2:31:03 PM
✘ [ERROR]   Error: memory access out of bounds 
```

The errors started August 24, soon after https://github.com/cloudflare/azul/commit/ad6ced9d2b0a7850230777b8299a47120640b330 was deployed, which fixes the issue reported in https://groups.google.com/a/chromium.org/g/ct-policy/c/dVz6tMLWsRY/m/ApNZBFkjGwAJ. We decided not to roll back as that would re-introduce the bug. We've been able to reproduce the issue for the cftest log shards, and confirmed that rolling back does fix the memory issues.

Closer inspection of the commit in question reveals that we keep the 'roots' variable (containing all accepted roots for the log) in scope from https://github.com/cloudflare/azul/commit/ad6ced9d2b0a7850230777b8299a47120640b330#diff-130a857a7156f3c81dd0e04249775b84bf4439aed5846bddaa27337e84ce06c4R195 all the way until the return at https://github.com/cloudflare/azul/commit/ad6ced9d2b0a7850230777b8299a47120640b330#diff-130a857a7156f3c81dd0e04249775b84bf4439aed5846bddaa27337e84ce06c4R311, instead of immediately releasing the variable from scope so that the memory is de-allocated. Specifically, the variable is kept in memory across the asynchronous call to the sequencer, which typically takes 1-2 seconds. Across many requests, memory is quickly exhausted, and DOs also seem to be impacted.

--- a/crates/ct_worker/Cargo.toml
+++ b/crates/ct_worker/Cargo.toml
@@ -51,6 +51,7 @@ serde_with.workspace = true
 static_ct_api.workspace = true
 signed_note.workspace = true
 tlog_tiles.workspace = true
+tokio.workspace = true
 worker.workspace = true
 x509-cert.workspace = true
 x509_util.workspace = true

--- a/crates/ct_worker/src/ccadb_roots_cron.rs
+++ b/crates/ct_worker/src/ccadb_roots_cron.rs
@@ -71,7 +71,7 @@ pub(crate) async fn update_ccadb_roots<T: AsRef<str>>(keys: &[T], kv: &KvStore) 
             buf = old;
         }
         for cert in &ccadb_roots {
-            if pool.included(cert).map_err(|e| e.to_string())? {
+            if pool.includes(cert).map_err(|e| e.to_string())? {
                 continue;
             }
             new_roots += 1;

--- a/crates/ct_worker/src/frontend_worker.rs
+++ b/crates/ct_worker/src/frontend_worker.rs
@@ -204,7 +204,7 @@ async fn add_chain_or_pre_chain(
 
     // Temporal interval dates prior to the Unix epoch are treated as the Unix epoch.
     let roots = load_roots(env, name).await?;
-    let pending_entry = match static_ct_api::partially_validate_chain(
+    let (pending_entry, found_root_idx) = match static_ct_api::partially_validate_chain(
         &req.chain,
         roots,
         Some(
@@ -251,17 +251,12 @@ async fn add_chain_or_pre_chain(
         .map(Vec::as_slice)
         .collect::<Vec<&[u8]>>();
 
-    // Make sure the root is persisted as well.
-    let root_hash = pending_entry
-        .chain_fingerprints
-        .last()
-        .ok_or("chain fingerprints is empty")?;
-    let Some(root_cert) = roots.by_fingerprint(root_hash) else {
-        return Err(format!("{name}: root hash for chain is not in accepted roots").into());
-    };
-    let root_bytes = root_cert.to_der().map_err(|e| e.to_string())?;
-    if !issuers.iter().any(|issuer| *issuer == root_bytes) {
-        issuers.push(root_bytes.as_slice());
+    // Make sure the inferred root is persisted as well, if the add-chain
+    // request did not include the root.
+    let root_bytes;
+    if let Some(idx) = found_root_idx {
+        root_bytes = roots.certs[idx].to_der().map_err(|e| e.to_string())?;
+        issuers.push(&root_bytes);
     }
 
     generic_log_worker::upload_issuers(&public_bucket, &issuers, name).await?;

--- a/crates/ct_worker/src/frontend_worker.rs
+++ b/crates/ct_worker/src/frontend_worker.rs
@@ -203,7 +203,7 @@ async fn add_chain_or_pre_chain(
     };
 
     // Temporal interval dates prior to the Unix epoch are treated as the Unix epoch.
-    let roots = &load_roots(env, name).await?;
+    let roots = load_roots(env, name).await?;
     let pending_entry = match static_ct_api::partially_validate_chain(
         &req.chain,
         roots,

--- a/crates/ct_worker/src/frontend_worker.rs
+++ b/crates/ct_worker/src/frontend_worker.rs
@@ -244,22 +244,24 @@ async fn add_chain_or_pre_chain(
 
     // Entry is not cached, so we need to sequence it.
 
-    // First persist issuers.
-    let public_bucket = ObjectBucket::new(load_public_bucket(env, name)?);
-    let mut issuers = req.chain[1..]
-        .iter()
-        .map(Vec::as_slice)
-        .collect::<Vec<&[u8]>>();
+    // First persist issuers. Use a block so memory is deallocated sooner.
+    {
+        let public_bucket = ObjectBucket::new(load_public_bucket(env, name)?);
+        let mut issuers = req.chain[1..]
+            .iter()
+            .map(Vec::as_slice)
+            .collect::<Vec<&[u8]>>();
 
-    // Make sure the inferred root is persisted as well, if the add-chain
-    // request did not include the root.
-    let root_bytes;
-    if let Some(idx) = found_root_idx {
-        root_bytes = roots.certs[idx].to_der().map_err(|e| e.to_string())?;
-        issuers.push(&root_bytes);
+        // Make sure the inferred root is persisted as well, if the add-chain
+        // request did not include the root.
+        let root_bytes;
+        if let Some(idx) = found_root_idx {
+            root_bytes = roots.certs[idx].to_der().map_err(|e| e.to_string())?;
+            issuers.push(&root_bytes);
+        }
+
+        generic_log_worker::upload_issuers(&public_bucket, &issuers, name).await?;
     }
-
-    generic_log_worker::upload_issuers(&public_bucket, &issuers, name).await?;
 
     // Submit entry to be sequenced, either via a batcher or directly to the
     // sequencer.

--- a/crates/ct_worker/src/frontend_worker.rs
+++ b/crates/ct_worker/src/frontend_worker.rs
@@ -252,7 +252,7 @@ async fn add_chain_or_pre_chain(
             .map(Vec::as_slice)
             .collect::<Vec<&[u8]>>();
 
-        // Make sure the inferred root is persisted as well, if the add-chain
+        // Make sure the found root is persisted as well, if the add-chain
         // request did not include the root.
         let root_bytes;
         if let Some(idx) = found_root_idx {

--- a/crates/ct_worker/src/lib.rs
+++ b/crates/ct_worker/src/lib.rs
@@ -11,6 +11,7 @@ use static_ct_api::StaticCTCheckpointSigner;
 use std::collections::HashMap;
 use std::sync::{LazyLock, OnceLock};
 use tlog_tiles::{CheckpointSigner, Ed25519CheckpointSigner, SequenceMetadata};
+use tokio::sync::OnceCell;
 use worker::{Env, Result};
 use x509_util::CertPool;
 
@@ -34,6 +35,7 @@ static CONFIG: LazyLock<AppConfig> = LazyLock::new(|| {
 
 static SIGNING_KEY_MAP: OnceLock<HashMap<String, OnceLock<EcdsaSigningKey>>> = OnceLock::new();
 static WITNESS_KEY_MAP: OnceLock<HashMap<String, OnceLock<Ed25519SigningKey>>> = OnceLock::new();
+static ROOTS: OnceCell<CertPool> = OnceCell::const_new();
 
 pub(crate) fn load_signing_key(env: &Env, name: &str) -> Result<&'static EcdsaSigningKey> {
     let once = &SIGNING_KEY_MAP.get_or_init(|| {
@@ -107,33 +109,37 @@ fn ccadb_roots_filename(name: &str) -> String {
     format!("roots_{name}.pem")
 }
 
-async fn load_roots(env: &Env, name: &str) -> Result<CertPool> {
+async fn load_roots(env: &Env, name: &str) -> Result<&'static CertPool> {
     // Load embedded roots.
-    let pem = include_bytes!(concat!(env!("OUT_DIR"), "/roots.pem"));
-    let mut pool = CertPool::default();
-    // load_pem_chain fails on empty input: https://github.com/RustCrypto/formats/pull/1965
-    if !pem.is_empty() {
-        pool.append_certs_from_pem(pem)
-            .map_err(|e| format!("failed to load PEM chain: {e}"))?;
-    }
+    ROOTS
+        .get_or_try_init(|| async {
+            let pem = include_bytes!(concat!(env!("OUT_DIR"), "/roots.pem"));
+            let mut pool = CertPool::default();
+            // load_pem_chain fails on empty input: https://github.com/RustCrypto/formats/pull/1965
+            if !pem.is_empty() {
+                pool.append_certs_from_pem(pem)
+                    .map_err(|e| format!("failed to load PEM chain: {e}"))?;
+            }
 
-    // Load additional roots from the CCADB roots file in Workers KV.
-    if CONFIG.logs[name].enable_ccadb_roots {
-        let key = ccadb_roots_filename(name);
-        let kv = env.kv(CCADB_ROOTS_NAMESPACE)?;
-        let pem = if let Some(pem) = kv.get(&key).text().await? {
-            pem
-        } else {
-            // The roots file might not exist if the CCADB roots cron job hasn't
-            // run yet. Try to create it once before failing.
-            update_ccadb_roots(&[&key], &kv).await?;
-            kv.get(&key)
-                .text()
-                .await?
-                .ok_or(format!("{name}: '{key}' not found in KV"))?
-        };
-        pool.append_certs_from_pem(pem.as_bytes())
-            .map_err(|e| format!("failed to add CCADB certs to pool: {e}"))?;
-    }
-    Ok(pool)
+            // Load additional roots from the CCADB roots file in Workers KV.
+            if CONFIG.logs[name].enable_ccadb_roots {
+                let key = ccadb_roots_filename(name);
+                let kv = env.kv(CCADB_ROOTS_NAMESPACE)?;
+                let pem = if let Some(pem) = kv.get(&key).text().await? {
+                    pem
+                } else {
+                    // The roots file might not exist if the CCADB roots cron job hasn't
+                    // run yet. Try to create it once before failing.
+                    update_ccadb_roots(&[&key], &kv).await?;
+                    kv.get(&key)
+                        .text()
+                        .await?
+                        .ok_or(format!("{name}: '{key}' not found in KV"))?
+                };
+                pool.append_certs_from_pem(pem.as_bytes())
+                    .map_err(|e| format!("failed to add CCADB certs to pool: {e}"))?;
+            }
+            Ok(pool)
+        })
+        .await
 }

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -176,21 +176,19 @@ pub fn partially_validate_chain(
     // The last certificate in the chain is either a root certificate
     // or a certificate that chains to a known root certificate.
     let mut found_root_idx = None;
-    let to_verify_issuer = to_verify.tbs_certificate.issuer.to_string();
     if !roots.includes(to_verify)? {
-        let mut found = false;
-        for &idx in roots.find_potential_parents(to_verify)? {
-            if is_link_valid(to_verify, &roots.certs[idx]) {
-                found_root_idx = Some(idx);
-                chain_certs.push(roots.certs[idx].clone());
-                chain_fingerprints.push(Sha256::digest(roots.certs[idx].to_der()?).into());
-                found = true;
-                break;
-            }
-        }
-        if !found {
-            return Err(StaticCTError::NoPathToTrustedRoot { to_verify_issuer });
-        }
+        let Some(&found_idx) = roots
+            .find_potential_parents(to_verify)?
+            .iter()
+            .find(|&&roots_idx| is_link_valid(to_verify, &roots.certs[roots_idx]))
+        else {
+            return Err(StaticCTError::NoPathToTrustedRoot {
+                to_verify_issuer: to_verify.tbs_certificate.issuer.to_string(),
+            });
+        };
+        found_root_idx = Some(found_idx);
+        chain_certs.push(roots.certs[found_idx].clone());
+        chain_fingerprints.push(Sha256::digest(&roots.certs[found_idx].to_der()?).into());
     }
 
     // Construct a pending log entry.

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -175,24 +175,22 @@ pub fn partially_validate_chain(
 
     // The last certificate in the chain is either a root certificate
     // or a certificate that chains to a known root certificate.
-    let mut found = false;
     let mut found_root_idx = None;
     let to_verify_issuer = to_verify.tbs_certificate.issuer.to_string();
-    for &idx in roots.find_potential_parents(to_verify)? {
-        if to_verify == &roots.certs[idx] {
-            found = true;
-            break;
+    if !roots.includes(to_verify)? {
+        let mut found = false;
+        for &idx in roots.find_potential_parents(to_verify)? {
+            if is_link_valid(to_verify, &roots.certs[idx]) {
+                found_root_idx = Some(idx);
+                chain_certs.push(roots.certs[idx].clone());
+                chain_fingerprints.push(Sha256::digest(roots.certs[idx].to_der()?).into());
+                found = true;
+                break;
+            }
         }
-        if is_link_valid(to_verify, &roots.certs[idx]) {
-            found_root_idx = Some(idx);
-            chain_certs.push(roots.certs[idx].clone());
-            chain_fingerprints.push(Sha256::digest(roots.certs[idx].to_der()?).into());
-            found = true;
-            break;
+        if !found {
+            return Err(StaticCTError::NoPathToTrustedRoot { to_verify_issuer });
         }
-    }
-    if !found {
-        return Err(StaticCTError::NoPathToTrustedRoot { to_verify_issuer });
     }
 
     // Construct a pending log entry.
@@ -569,6 +567,10 @@ mod tests {
     test_validate_chain!(missing_server_auth_eku_not_required; "../tests/fake-root-ca-cert.pem"; "../tests/subleaf.chain"; None; None; false; false; false; 3);
     test_validate_chain!(missing_server_auth_eku_required; "../tests/fake-root-ca-cert.pem"; "../tests/subleaf.chain"; None; None; false; true; true; 0);
     test_validate_chain!(preissuer_chain; "../tests/test-roots.pem"; "../tests/preissuer-chain.pem"; None; None; true; true; false; 3);
+
+    test_validate_chain!(intermediate_as_accepted_root; "../tests/fake-intermediate-cert.pem"; "../tests/leaf-signed-by-fake-intermediate-cert.pem"; None; None; false; true; false; 1);
+
+    test_validate_chain!(leaf_as_accepted_root; "../tests/leaf-signed-by-fake-intermediate-cert.pem"; "../tests/leaf-signed-by-fake-intermediate-cert.pem"; None; None; false; true; false; 0);
 
     #[test]
     fn test_build_precert_tbs() {

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -159,12 +159,12 @@ pub fn partially_validate_chain(
             has_precert_signing_cert = true;
         }
 
-        // Also check that intermediates have the CA Basic Constraint,
-        // but don't enforce for Precertificate Signing Certificates.
+        // Check that intermediates have the CA Basic Constraint.
+        // Precertificate signing certificates must also have CA:true.
         if cert
             .tbs_certificate
             .get::<BasicConstraints>()?
-            .is_some_and(|(_, bc)| !bc.ca)
+            .is_none_or(|(_, bc)| !bc.ca)
         {
             return Err(StaticCTError::IntermediateMissingCABasicConstraint);
         }

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -134,38 +134,43 @@ pub fn partially_validate_chain(
     // We can now do the verification. Use fairly lax options for verification, as
     // CT is intended to observe certifications rather than police them.
 
-    let mut intermediates: Vec<Certificate> = raw_chain[1..]
+    let mut chain_certs: Vec<Certificate> = raw_chain[1..]
         .iter()
         .map(|v| parse_certificate(v))
         .collect::<Result<_, _>>()?;
 
     let mut chain_fingerprints: Vec<[u8; 32]> = raw_chain[1..]
         .iter()
-        .map(|issuer| Sha256::digest(issuer).into())
+        .map(|v| Sha256::digest(v).into())
         .collect();
 
     // Walk up the chain, ensuring that each certificate signs the previous one.
     // This simplified chain validation is possible due to the constraints laid out in RFC 6962.
     let mut to_verify = &leaf;
     let mut has_precert_signing_cert = false;
-    for (idx, ca) in intermediates.iter().enumerate() {
-        if !is_link_valid(to_verify, ca) {
+    for (idx, cert) in chain_certs.iter().enumerate() {
+        // Check that this cert signs the previous one in the chain.
+        if !is_link_valid(to_verify, cert) {
             return Err(StaticCTError::InvalidLinkInChain);
         }
-        to_verify = ca;
+
+        // Record if this is a precertificate signing certificate.
+        if is_precert && idx == 0 && is_pre_issuer(cert)? {
+            has_precert_signing_cert = true;
+        }
 
         // Also check that intermediates have the CA Basic Constraint,
         // but don't enforce for Precertificate Signing Certificates.
-        if is_precert && idx == 0 && is_pre_issuer(ca)? {
-            has_precert_signing_cert = true;
-        }
-        if ca
+        if cert
             .tbs_certificate
             .get::<BasicConstraints>()?
             .is_some_and(|(_, bc)| !bc.ca)
         {
             return Err(StaticCTError::IntermediateMissingCABasicConstraint);
         }
+
+        // Verify this cert next.
+        to_verify = cert;
     }
 
     // The last certificate in the chain is either a root certificate
@@ -180,7 +185,7 @@ pub fn partially_validate_chain(
         }
         if is_link_valid(to_verify, &roots.certs[idx]) {
             found_root_idx = Some(idx);
-            intermediates.push(roots.certs[idx].clone());
+            chain_certs.push(roots.certs[idx].clone());
             chain_fingerprints.push(Sha256::digest(roots.certs[idx].to_der()?).into());
             found = true;
             break;
@@ -197,12 +202,12 @@ pub fn partially_validate_chain(
         let mut pre_issuer: Option<&TbsCertificate> = None;
         let issuer_key_hash: [u8; 32];
         if has_precert_signing_cert {
-            pre_issuer = Some(&intermediates[0].tbs_certificate);
-            if intermediates.len() < 2 {
+            pre_issuer = Some(&chain_certs[0].tbs_certificate);
+            if chain_certs.len() < 2 {
                 return Err(StaticCTError::MissingPrecertSigningCertificateIssuer);
             }
             issuer_key_hash = Sha256::digest(
-                intermediates[1]
+                chain_certs[1]
                     .tbs_certificate
                     .subject_public_key_info
                     .to_der()?,
@@ -210,7 +215,7 @@ pub fn partially_validate_chain(
             .into();
         } else {
             issuer_key_hash = Sha256::digest(
-                intermediates[0]
+                chain_certs[0]
                     .tbs_certificate
                     .subject_public_key_info
                     .to_der()?,

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -134,10 +134,14 @@ pub fn partially_validate_chain(
     // We can now do the verification. Use fairly lax options for verification, as
     // CT is intended to observe certifications rather than police them.
 
-    let mut chain_certs: Vec<Certificate> = raw_chain[1..]
+    // Keep an owned copy of the parsed certs, but use a vector of references so
+    // we can avoid needing to clone the inferred root cert when adding it to
+    // the chain.
+    let chain_certs_owned: Vec<Certificate> = raw_chain[1..]
         .iter()
         .map(|v| parse_certificate(v))
         .collect::<Result<_, _>>()?;
+    let mut chain_certs = chain_certs_owned.iter().collect::<Vec<_>>();
 
     let mut chain_fingerprints: Vec<[u8; 32]> = raw_chain[1..]
         .iter()
@@ -187,7 +191,7 @@ pub fn partially_validate_chain(
             });
         };
         found_root_idx = Some(found_idx);
-        chain_certs.push(roots.certs[found_idx].clone());
+        chain_certs.push(&roots.certs[found_idx]);
         chain_fingerprints.push(Sha256::digest(&roots.certs[found_idx].to_der()?).into());
     }
 

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -94,11 +94,10 @@ pub fn partially_validate_chain(
     require_server_auth_eku: bool,
 ) -> Result<(StaticCTPendingLogEntry, Option<usize>), StaticCTError> {
     // First make sure the cert is well-formed.
-    let mut iter = raw_chain.iter();
-    let leaf: Certificate = match iter.next() {
-        Some(v) => parse_certificate(v)?,
-        None => return Err(StaticCTError::EmptyChain),
-    };
+    if raw_chain.is_empty() {
+        return Err(StaticCTError::EmptyChain);
+    }
+    let leaf = parse_certificate(&raw_chain[0])?;
 
     // Check whether the expiry date is within the acceptable range for this log shard.
     let not_after = u64::try_from(
@@ -135,7 +134,8 @@ pub fn partially_validate_chain(
     // We can now do the verification. Use fairly lax options for verification, as
     // CT is intended to observe certifications rather than police them.
 
-    let mut intermediates: Vec<Certificate> = iter
+    let mut intermediates: Vec<Certificate> = raw_chain[1..]
+        .iter()
         .map(|v| parse_certificate(v))
         .collect::<Result<_, _>>()?;
 

--- a/crates/static_ct_api/src/rfc6962.rs
+++ b/crates/static_ct_api/src/rfc6962.rs
@@ -92,7 +92,7 @@ pub fn partially_validate_chain(
     not_after_end: Option<UnixTimestamp>,
     expect_precert: bool,
     require_server_auth_eku: bool,
-) -> Result<StaticCTPendingLogEntry, StaticCTError> {
+) -> Result<(StaticCTPendingLogEntry, Option<usize>), StaticCTError> {
     // First make sure the cert is well-formed.
     let mut iter = raw_chain.iter();
     let leaf: Certificate = match iter.next() {
@@ -171,6 +171,7 @@ pub fn partially_validate_chain(
     // The last certificate in the chain is either a root certificate
     // or a certificate that chains to a known root certificate.
     let mut found = false;
+    let mut found_root_idx = None;
     let to_verify_issuer = to_verify.tbs_certificate.issuer.to_string();
     for &idx in roots.find_potential_parents(to_verify)? {
         if to_verify == &roots.certs[idx] {
@@ -178,6 +179,7 @@ pub fn partially_validate_chain(
             break;
         }
         if is_link_valid(to_verify, &roots.certs[idx]) {
+            found_root_idx = Some(idx);
             intermediates.push(roots.certs[idx].clone());
             chain_fingerprints.push(Sha256::digest(roots.certs[idx].to_der()?).into());
             found = true;
@@ -226,11 +228,14 @@ pub fn partially_validate_chain(
         certificate = leaf.to_der()?;
     }
 
-    Ok(StaticCTPendingLogEntry {
-        certificate,
-        precert_opt,
-        chain_fingerprints,
-    })
+    Ok((
+        StaticCTPendingLogEntry {
+            certificate,
+            precert_opt,
+            chain_fingerprints,
+        },
+        found_root_idx,
+    ))
 }
 
 /// Precertificate poison extension that can be decoded with [`TbsCertificate::get`].
@@ -462,7 +467,7 @@ mod tests {
                     );
                 assert_eq!(result.is_err(), $want_err);
 
-                if let Ok(pending_entry) = result {
+                if let Ok((pending_entry, _found_root_idx)) = result {
                     assert_eq!(pending_entry.chain_fingerprints.len(), $want_chain_len);
                 }
             }

--- a/crates/x509_util/src/lib.rs
+++ b/crates/x509_util/src/lib.rs
@@ -52,7 +52,7 @@ impl CertPool {
         Ok(pool)
     }
 
-    /// Search the certificate pool for potential parents for the provided certificates.
+    /// Search the certificate pool for potential parents for the provided certificate.
     ///
     /// # Errors
     ///
@@ -109,12 +109,12 @@ impl CertPool {
         Ok(())
     }
 
-    /// Check if a certificate is included in the pool.
+    /// Check if the pool includes a certificate.
     ///
     /// # Errors
     ///
     /// Returns an error if there are issues DER-encoding the certificate.
-    pub fn included(&self, cert: &Certificate) -> Result<bool, DerError> {
+    pub fn includes(&self, cert: &Certificate) -> Result<bool, DerError> {
         Ok(self
             .by_fingerprint
             .contains_key::<[u8; 32]>(&Sha256::digest(cert.to_der()?).into()))

